templates/fine-tune-llm_v2/cookbooks/optimize_cost/README.md

```diff
@@ -0,0 +1,232 @@
+
+A prime example is running fine-tuning on `A10G`s. On AWS `A10s` are available in `g5.4xlarge` with 1 GPU and small RAM capacity all the way to `g5.48xlarge` with 8 GPUs and large RAM. During checkpoining a large model like 70B, the CPU RAM on g5.4xlarge is not sufficient and hence we have to define `memory` requirement for `trained_resources` to ensure that the large instance gets picked for rank-0. For example let's look at `./training_configs/full_param/llama-3-70b.yaml`:
+
+```
+num_devices: 32 
+# Head node would have at least 200 GB memory
+trainer_resources:
+  memory: 161061273600 # 130 GB memory
+worker_resources:
+  memory: 53687091200 # 50 GB memory
+  accelerator_type:A10G: 0.001
+```
+
+This configs is asking for 32xA10G GPUs but does not specify the architecture (e.g. `g5.4xlarge` vs. `g5.48xlarge`). However, it specifies that there should be at least 130G of memory available for rank-0, which forces the Anyscale's instance manager to pick g5.48xlarge for rank-0 and any other A10s for other workers hence not hitting RAM OOM issues during checkpointing. This is much better than being forced to use `4xg5.48xlarge` nodes. They may not be available on-demand and might be more expensive too.
+
+## Batch size and Context length
+
+Usually for large models we want to saturate the GPU's Memory by choosing the largest batch size before we OOM. If you tried to fit a micro batch size of 4 and it OOMed, but 3 fits instead of 2. Because, sequence length dimensions are mostly flattened for linear layer computation the batch size for batch matrix-multiplication is sequence_length x batch_size_per_device. So the GPUs would remain in their high-performant regime even if batch size is seemingly small.
+
+To increase batch size beyond the capacity of GPUs, you either need to user more instances, or use [gradient accumulation]() which may be a better option when the difference is between multi-node vs. single node.
+
+If batch size or context length are not specified they are automatically inferred based on some heuristics that may not be optimals of your use-case. Batch size is simply chosen based on a look up table of previous runs based on the context length. The context length is chose to be the max. of model's default context length and 95th precentile of the sequence length of your dataset.
+
+To change these configs you can simply do:
+
+```
+context_length: <ctx_length>
+train_batch_size_per_device: <bs>
+# You can also change the validation batch size
+eval_batch_size_per_device: <bs>
+```
+
+## Deepspeed configs
+
+By default, we encourage everyone to use Zero-DP stage 3 with parameter and optimizare's state offloading to be the safest in terms of memory consumption. But this setting is not the fastest. You can find other typical deepseepd configurations under `deepspeed_Configs`. A complete doc on all the configurations can be found on [deepspeeds doc page](https://www.deepspeed.ai/docs/config-json/).
```

Comment:
```suggestion
By default, we encourage everyone to use Zero-DP stage 3 with parameter and optimizare's state offloading to be the safest in terms of memory consumption. But this setting is not the fastest. You can find other typical deepspeed configurations under `deepspeed_Configs`. A complete doc on all the configurations can be found on [deepspeeds doc page](https://www.deepspeed.ai/docs/config-json/).
```

Reply:
Is there a reason for C to be capitalized in deepspeed_Configs?

Reply:
No. typo.